pkg/image/admission/imagepolicy/api/types.go

```diff
@@ -0,0 +1,74 @@
+package api
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+// IgnorePolicyRulesAnnotation is a comma delimited list of rule names to omit from consideration
+// in a given namespace. Loaded from the namespace.
+const IgnorePolicyRulesAnnotation = "alpha.image.policy.openshift.io/ignore-rules"
+
+// ImagePolicyConfig is the configuration for controlling how images are used in the cluster.
+type ImagePolicyConfig struct {
+	unversioned.TypeMeta
+
+	// ExecutionRules determine whether the use of an image is allowed in an object with a pod spec.
+	// By default, these rules only apply to pods, but may be extended to other resource types.
+	ExecutionRules []ImageExecutionPolicyRule
+}
+
+// ImageExecutionPolicyRule determines whether a provided image may be used on the platform.
+type ImageExecutionPolicyRule struct {
+	ImageCondition
+
+	// Resolve indicates that images referenced by this resource must be resolved
+	Resolve bool
+}
+
+// ImageCondition defines the conditions for matching a particular image source. The conditions below
+// are all required (logical AND). If Reject is specified, the condition is false if all conditions match,
+// and true otherwise.
+type ImageCondition struct {
+	// Name is the name of this policy rule for reference. It must be unique across all rules.
+	Name string
+	// IgnoreNamespaceOverride prevents this condition from being overriden when the
+	// `alpha.image.policy.openshift.io/ignore-rules` is set on a namespace and contains this rule name.
+	IgnoreNamespaceOverride bool
+
+	// Reject indicates this rule is inverted - if all conditions below match, then this rule is considered to
+	// not match.
+	Reject bool
+
+	// OnResources determines which resources this applies to. Defaults to 'pods' for ImageExecutionPolicyRules.
+	OnResources []unversioned.GroupResource
+
+	// MatchIntegratedRegistry will only match image sources that originate from the configured integrated
+	// registry.
+	MatchIntegratedRegistry bool
+	// MatchRegistries will match image references that point to the provided registries
+	MatchRegistries []string
+
+	// AllowResolutionFailure allows the subsequent conditions to be bypassed if the integrated registry does
+	// not have access to image metadata (no image exists matching the image digest).
+	AllowResolutionFailure bool
+
+	// MatchDockerImageLabels checks against the resolved image for the presence of a Docker label
+	MatchDockerImageLabels []ValueCondition
+	// MatchImageLabels checks against the resolved image for a label.
+	MatchImageLabels []unversioned.LabelSelector
+	// MatchImageLabelSelectors is the processed form of MatchImageLabels
+	MatchImageLabelSelectors []labels.Selector
```

Comment:
I believe we ended up using the "efficient runtime structure" in the
internal object of cluster resource quota?

On Tue, Aug 9, 2016 at 9:16 AM, David Eads <notifications@github.com> wrote:

> In pkg/image/admission/imagepolicy/api/types.go
> <https://github.com/openshift/origin/pull/8995#discussion_r74055037>:
>
> > +	// MatchIntegratedRegistry will only match image sources that originate from the configured integrated
> > +	// registry.
> > +	MatchIntegratedRegistry bool
> > +	// MatchRegistries will match image references that point to the provided registries
> > +	MatchRegistries []string
> > +
> > +	// AllowResolutionFailure allows the subsequent conditions to be bypassed if the integrated registry does
> > +	// not have access to image metadata (no image exists matching the image digest).
> > +	AllowResolutionFailure bool
> > +
> > +	// MatchDockerImageLabels checks against the resolved image for the presence of a Docker label
> > +	MatchDockerImageLabels []ValueCondition
> > +	// MatchImageLabels checks against the resolved image for a label.
> > +	MatchImageLabels []unversioned.LabelSelector
> > +	// MatchImageLabelSelectors is the processed form of MatchImageLabels
> > +	MatchImageLabelSelectors []labels.Selector
>
> This is an AND condition, and multiple selectors can't be cleanly composed
> with the code we had.
>
> This is *exactly* what you've smacked me for before in authorization.
> This is supposed to be the data, the efficient runtime structure is
> supposed to be separate.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/8995/files/9bbe455e51d9ecc555a3089d8911ba8da6c42c0c#r74055037>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwOA_nDHsoE_ZsL65UHG5vlyb22Bks5qeH2mgaJpZM4IlDlb>
> .
>